File: Sources/stub.swift
//
//  KVO.stub.swift
//  KVO.swift
//
//

import Foundation
import ObjectiveC.runtime

// Since the fucntion pointer is not available in Swift, we use @convention(c) to stub the setter
// instead of Swift func
// (https://stackoverflow.com/questions/43880839/swift-unable-to-cast-function-pointer-to-void-for-use-in-c-style-third-party)

typealias _setter_stub_impl_type = @convention(c) (_ self: NSObject , _ _cmd: Selector, _ newValue: Any) -> Void

@_silgen_name("objc_msgSendSuper") func objc_msgSendSuper(_: UnsafeMutablePointer<objc_super>, _: Selector, _: Any) -> Void

@_silgen_name("objc_msgSend") func objc_msgSend(_: AnyObject, _: Selector, _: Any) -> Void

@_silgen_name("object_getSuper") func object_getSuper(_: AnyObject) -> objc_super

let _setter_stub_impl: _setter_stub_impl_type = { (`self`, _cmd, newValue) -> Void in
    let sel = _cmd.getter()
    let key = NSStringFromSelector(sel!)
    let cls: AnyClass = object_getClass(self)!
    
    guard let _ = class_getInstanceMethod(cls, sel) else {
        // FIXME: throw error
        return
    }
    let oldValue = self.value(forKey: key)!
    
    // FIXME: fire a bug
    // Command failed due to signal: Abort trap: 6
    // IntToPtr source must be an integral
    //    %4 = inttoptr %swift.type* %1 to %objc_class*, !dbg !178
    // var _spr = objc_super(receiver: Unmanaged.passUnretained(`self`),
    //                       åsuper_class: class_getSuperclass(object_getClass(`self`)))
    // use c function bridge
    var spr = object_getSuper(`self`)
    
    var ctx: Observable.Context =
        objc_getAssociatedObject(self, Observable.contextKey) as? Observable.Context ?? [
            Observable.willSet: [],
            Observable.didSet: []
        ]

    if let willSet = ctx[Observable.willSet] {
        willSet.forEach({ (body) in
            (body as? ((Any) -> Void))?(newValue)
        })
    }
  
    // @objc thunk lookUpImpOrForward crash
    // objc_msgSendSuper(&spr, _cmd, newValue)
    
    if let didSet = ctx[Observable.didSet] {
        didSet.forEach({ (body) in
            (body as? ((Any) -> Void))?(oldValue)
        })
    }
}

let _class_impl_stub: @convention (c) (_ self: AnyClass, _ _cmd: Selector) -> Void = { (`self`, _cmd) -> Void in
    return class_getSuperclass(object_getClass(self));
}

File: Sources/stub.c
//
//  stub.c
//  kvo
//
//

#include "stub.h"
#include "selproc.h"
#include "objc.h"
#include "obsv.h"
#include "obsv_private.h"
#include <objc/runtime.h>
#include <objc/message.h>
#include <objc/hashtable2.h>

void foreach(NXHashTable *table, void (^body)(id)) {
    if (!table) return;
    __auto_type count = 0;
    __auto_type buffer = nil;
    __auto_type state = NXInitHashState(table);
    while (NXNextHashState(table, &state, &buffer)) {
        !body ?: body(buffer);
        count++;
    }
}

Class _class_impl_stub(id self, SEL _cmd) {
    return class_getSuperclass(object_getClass(self));
}

void _setter_impl_stub(id self, SEL _cmd, id newValue) {
    __auto_type getter = sel_getter(_cmd);
    __auto_type class = object_getClass(self);
    __auto_type method = class_getInstanceMethod(class, getter);
    if (!method)
        return;
    __auto_type oldValue = objc_msgSend(self, getter);
    __auto_type super = object_getSuper(self);
    
    foreach((NXHashTable *)objc_getAssociatedObject(self, &OBS_PRE_KEY), ^(id body) {
        !body ?: ((Callback)body)(newValue);
    });
    objc_msgSendSuper(&super, _cmd, newValue);
    
    foreach((NXHashTable *)objc_getAssociatedObject(self, &OBS_POST_KEY), ^(id body) {
        !body ?: ((Callback)body)(oldValue);
    });
}
